workspaces/src/types.rs

```diff
@@ -0,0 +1,243 @@
+/// Types copied over from near_primitives since those APIs are not yet stable.
+/// and internal libraries like near-jsonrpc-client requires specific versions
+/// of these types which shouldn't be exposed either.
+use std::convert::{TryFrom, TryInto};
+use std::fmt;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::path::Path;
+use std::str::FromStr;
+
+use near_crypto::key_conversion::convert_secret_key;
+use near_crypto::vrf::{Proof, Value};
+use near_crypto::Signature;
+use near_primitives::account::id::{MAX_ACCOUNT_ID_LEN, MIN_ACCOUNT_ID_LEN};
+
+pub(crate) use near_crypto::{KeyType, PublicKey, SecretKey, Signer};
+use serde::{Deserialize, Serialize};
+
+#[derive(Eq, Ord, Hash, Clone, Debug, PartialEq, PartialOrd, Serialize, Deserialize)]
+pub struct AccountId(Box<str>);
+
+impl AccountId {
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    pub fn validate(account_id: &str) -> Result<(), ParseAccountError> {
+        if account_id.len() < MIN_ACCOUNT_ID_LEN {
+            return Err(ParseAccountError(
+                ParseErrorKind::TooShort,
+                account_id.to_string(),
+            ));
+        }
+
+        if account_id.len() > MAX_ACCOUNT_ID_LEN {
+            return Err(ParseAccountError(
+                ParseErrorKind::TooLong,
+                account_id.to_string(),
+            ));
+        }
+        // Adapted from https://github.com/near/near-sdk-rs/blob/fd7d4f82d0dfd15f824a1cf110e552e940ea9073/near-sdk/src/environment/env.rs#L819
+
+        // NOTE: We don't want to use Regex here, because it requires extra time to compile it.
+        // The valid account ID regex is /^(([a-z\d]+[-_])*[a-z\d]+\.)*([a-z\d]+[-_])*[a-z\d]+$/
+        // Instead the implementation is based on the previous character checks.
+
+        // We can safely assume that last char was a separator.
+        let mut last_char_is_separator = true;
+
+        for c in account_id.bytes() {
+            let current_char_is_separator = match c {
+                b'a'..=b'z' | b'0'..=b'9' => false,
+                b'-' | b'_' | b'.' => true,
+                _ => {
+                    return Err(ParseAccountError(
+                        ParseErrorKind::Invalid,
+                        account_id.to_string(),
+                    ))
+                }
+            };
+            if current_char_is_separator && last_char_is_separator {
+                return Err(ParseAccountError(
+                    ParseErrorKind::Invalid,
+                    account_id.to_string(),
+                ));
+            }
+            last_char_is_separator = current_char_is_separator;
+        }
+
+        (!last_char_is_separator)
+            .then(|| ())
+            .ok_or_else(|| ParseAccountError(ParseErrorKind::Invalid, account_id.to_string()))
+    }
+}
```

Comment:
It exposes a near_primitive error though. What's the general strategy for exposing those? I think I could map that to the version here instead of duplicating all the validate logic

Reply:
`AccountId` specifically might be fine to expose. I got the feeling that we might publish 1.0 of it relatively soon cc @miraclx

Reply:
Yes indeed! `near-account-id 0.11.0` would be released later this week, sporting all the refactoring that's gone into it the past week. Also, while `near-account-id` is exposed from `near-primitives`, it's also its own [crate](https://docs.rs/near-account-id). So it's fine to expose.

Reply:
Well, I'd say that 0.11 is not super great to expose -- it implies that 0.12 would be released some time in the future, and if we expose it publicly here, that means that we'd have to publish a major as well, and *that* means that all the people who write tests for contracts need to upgrade to the new major version. When we release 1.0, we should make sure that everything we expose is also `1.0`. 

I guess let's do this -- expose AccountId directly for now while iterating in 0.1 ourselves. Once we publish 1.0, either wrap it in our own type or (preferred) pressure miraclx into releasing 1.0 as well :)

Reply:
To be clear, my suggestion was to keep the method as is, just re-use the internal impl and convert the error type to our own. Maybe I'm missing something so I will play around with this when I get to my computer

Reply:
hmm, I don't have an opinion on this, so I'll just get rid of it for now and expose the near_primitive one. Let me know @austinabell if you feel differently

Reply:
>and expose the near_primitive one

I belive you can use one from crates.io https://crates.io/crates/near-account-id. You might need the corresponding other crates from crates as well: 

* https://crates.io/crates/near-crypto
* https://crates.io/crates/near-primitives


Doing `pub use near_account_id::AccountId` is better than `pub use near_primitives::type::AccountId`. Although these are the same type right now, `near_primitives` does not guarantee that that'll be the case in the future as well.

Reply:
Got it, switched it over to `near-account-id` instead